### PR TITLE
ISPN-5625 Distributed stream iterator should try to stay local if pos…

### PR DESCRIPTION
**Pull request by gustavonalle**
…sible

https://issues.jboss.org/browse/ISPN-5625

Measured improvement of 2~3x in the Spark job execution (which relies heavily on iteration)

### CODE REVIEW

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 1]
```diff
@@ -431,104 +431,37 @@ public long count() {
...
+      CloseableIterator<R> closeableIterator = new CloseableSuppliedIterator<>(supplier);
+      onClose(() -> supplier.close());
+      return closeableIterator;
+   }
 
-         Thread thread = Thread.currentThread();
-         executor.execute(() -> {
-            try {
-               log.tracef("Thread %s submitted iterator request for stream", thread);
+   private void ignoreRehashIteration(Consumer<R> consumer, IteratorSupplier<R> supplier, boolean iteratorParallelDistribute) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Thanks, I was going to do this in another PR even :+1:

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 2]
```diff
@@ -537,23 +470,130 @@ public long count() {
...
+         }
+      });
+   }
 
-      CloseableIterator<R> closeableIterator = new CloseableSuppliedIterator<>(supplier);
-      onClose(() -> supplier.close());
-      return closeableIterator;
+   private Set<Integer> segmentsToProcess(IteratorSupplier<R> supplier, KeyTrackingConsumer<Object, R> results,
+                                          Set<Integer> segmentsToProcess, UUID id) {
+      String strId = id == null ? "local" : "identifier " + id;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: I would like if possible to avoid the string concat, maybe just leave it as id or some way of using ``if (log.traceEnabled())`` ?

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 3]
```diff
@@ -537,23 +470,130 @@ public long count() {
...
+                     } catch (InterruptedException e) {
+                        throw new CacheException(e);
+                     }
+                     segmentsToProcess = segmentsToProcess(supplier, results, segmentsToProcess, id);
+                  } finally {
+                     csm.forgetOperation(id);
+                  }
+               } else {
+                  performLocalRehashAwareOperation(results, segmentsToProcess, ch, segments, op,
+                          () -> ch.getSegmentsForOwner(localAddress), null);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: We can just return the segmentsForOwner from https://github.com/infinispan/infinispan/pull/3614/files#diff-be6f435a8edfd997f62d9a90ad59582bR514 if you push the variable out of the if brackets.  It can be defined as null and then only set if runLocal was set since we know stayLocal can only be true if it went into runLocal.  This will avoid an additional segmentsForOwner lookup.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 4]
```diff
@@ -537,23 +470,130 @@ public long count() {
...
+                       distributedBatchSize);
+               if (!stayLocal) {
+                  UUID id = csm.remoteStreamOperationRehashAware(iteratorParallelDistribute, parallel, ch,
+                          segmentsToProcess, keysToFilter, new AtomicReferenceArrayToMap<>(results.referenceArray),
+                          includeLoader, op, results);
+                  supplier.pending = id;
+                  try {
+                     if (runLocal) {
+                        performLocalRehashAwareOperation(results, segmentsToProcess, ch, segments, op,
+                                () -> ch.getPrimarySegmentsForOwner(localAddress), id);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Same here with getPrimarySegmentsForOwner, you can just pass segments as is since the way the logic works it will always be the primary owners.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 3 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Actually ignore what I said :)

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 4 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Actually ignore what I said :)

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorTest.java [hunk 5]
```diff
@@ -293,6 +296,44 @@ public void testLocallyForcedStream() {
...
+
+      ClusterStreamManager clusterStreamManager = replaceWithSpy(cache0);
+
+      IntStream.rangeClosed(0, 499).boxed().forEach(i -> cache0.put(i, i.toString()));
+
+      ConsistentHash ch = cache0.getAdvancedCache().getDistributionManager().getConsistentHash();
+      Set<Integer> segmentsCache0 = ch.getSegmentsForOwner(cache0.getCacheManager().getAddress());
+
+      CacheStream<Map.Entry<Object, String>> stream = cache0.entrySet().stream();
+      if (!rehashAware) stream.disableRehashAware();
```
Comment (wburns) [inline core/src/test/java/org/infinispan/stream/DistributedStreamIteratorTest.java]: We should reassign stream to the return value.

**Comment by wburns** [comment]
@gustavonalle Looks good, just a couple very minor comments.

Also a tip (not sure if you know or not), but when you change up the distributed cache stream code related to rehash, I would highly highly suggest running the DiistributedStreamRehashStressTest to verify there was no regression there.  Unfortunately it is a stress test so it doesn't get run with all of the other tests.

**Comment by gustavonalle** [comment]
@wburns PR updated, and ```DistributedStreamRehashStressTest``` is passing

**Comment by wburns** [comment]
Pulling...

**Comment by wburns** [comment]
Integrated into master, Thanks @gustavonalle !
